File: src/test/java/com/letgo/book/unit/application/create/CreateBookCommandHandlerTest.java
package com.letgo.book.unit.application.create;

import com.letgo.book.application.create.CreateBookCommand;
import com.letgo.book.application.create.CreateBookCommandHandler;
import com.letgo.book.domain.Book;
import com.letgo.book.domain.BookCreated;
import com.letgo.book.domain.BookId;
import com.letgo.book.domain.BookTitle;
import com.letgo.book.unit.application.BookTestCase;
import com.letgo.book.unit.domain.BookIdMother;
import com.letgo.book.unit.domain.BookTitleMother;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertEquals;

final public class CreateBookCommandHandlerTest extends BookTestCase {
    private final CreateBookCommandHandler handler = new CreateBookCommandHandler(repository, publisher);

    @Test
    public void itShouldCreateABook() {
        BookId id = BookIdMother.random();
        BookTitle title = BookTitleMother.create("Title");

        expectDomainEventsToBePublished(new BookCreated(id.value(), title.value()));

        handler.handle(new CreateBookCommand(id.value(), title.value()));

        Book book = repository.find(id).orElseThrow();
        assertEquals(id, book.id());
        assertEquals(title, book.title());

        eventsShouldBePublished();
    }

    @Test
    public void itShouldBeIdempotent() {
        Book currentBook = anExistingBook();
        BookId id = currentBook.id();

        expectDomainEventsToBePublished();

        handler.handle(new CreateBookCommand(id.value(), currentBook.title().value()));

        assertEquals(currentBook, repository.find(id).orElseThrow());

        eventsShouldNotBePublished();
    }
}

File: src/test/java/com/letgo/book/unit/application/create/CreateBookCommandHandlerTest.kt
package com.letgo.book.unit.application.create

import com.letgo.book.application.create.CreateBookCommand
import com.letgo.book.application.create.CreateBookCommandHandler
import com.letgo.book.domain.BookCreated
import com.letgo.book.unit.application.BookTestCase
import com.letgo.book.unit.domain.ABook
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test

private class CreateBookCommandHandlerTest : BookTestCase() {
    private val handler = CreateBookCommandHandler(repository, publisher)

    @Test
    fun `It should create a book`() {
        val expectedBook = ABook.random()
        val id = expectedBook.id()
        val title = expectedBook.title()

        expectDomainEventsToBePublished(
            BookCreated(
                aggregateId = id.value(),
                title = title.value(),
                occurredOn = title.createdAt(),
            )
        )

        handler.handle(
            CreateBookCommand(
                id = id.value(),
                title = title.value(),
                occurredOn = title.createdAt().toString(),
            )
        )

        assertEquals(expectedBook, repository.find(id))

        eventsShouldBePublished()
    }

    @Test
    fun `It should be idempotent`() {
        val currentBook = anExistingBook()
        val currentBookId = currentBook.id()
        val currentTitle = currentBook.title()

        expectDomainEventsToBePublished()

        handler.handle(
            CreateBookCommand(
                id = currentBookId.value(),
                title = currentTitle.value(),
                occurredOn = currentTitle.createdAt().toString(),
            )
        )

        assertEquals(currentBook, repository.find(currentBookId))

        eventsShouldNotBePublished()
    }
}
